lib/active_merchant/billing/gateways/shift4.rb

```diff
@@ -0,0 +1,277 @@
+      end
+
+      def add_p2pe(post, options)
+        if options[:p2pe].present?
+          post[:p2pe] = {}
+          post[:p2pe][:format] = options[:p2pe][:format]
+          post[:p2pe][:ksn] = options[:p2pe][:ksn]
+        end
+      end
+
+      def add_card(post, credit_card, options)
+        post[:card] = {}
+        if options[:token].present?
+          post[:card][:token] = {}
+          post[:card][:token][:value] = options[:token][:value]
+        else
+          if credit_card.respond_to?(:number)
+            post[:card][:expirationDate] = "0#{credit_card.month}#{credit_card.year.to_s[2..4]}"
+            post[:card][:number] = credit_card.number
+            post[:card][:entryMode] = options[:entry_mode]
+            post[:card][:present] = options[:present]
+            post[:card][:securityCode] = {}
+            if options[:security_code].present?
+              post[:card][:securityCode][:indicator] = options[:security_code][:indicator]
+              post[:card][:securityCode][:value] = options[:security_code][:value]
+            end
+          end
+          end
+
+      end
+
+      def add_customer(post, options)
+        if address = options[:customer]
+          post[:customer] = {}
+          post[:customer][:addressLine1] = address[:address_line1] if address[:address_line1]
+          post[:customer][:firstName] = address[:first_name]
+          post[:customer][:lastName] = address[:last_name]
+          post[:customer][:postalCode] = address[:postal_code]
+        end
+      end
+
+      def commit(action, parameters, headers)
+        url_postfix = action == 'accesstoken' ? 'credentials' : 'transactions'
+        url = (test? ? "#{test_url}#{url_postfix}/#{action}" : "#{live_url}#{url_postfix}/#{action}")
+        response = parse(ssl_post(url, parameters.to_json, headers))
+        Response.new(
+          success_from(response),
+          message_from(response),
+          response,
+          authorization: authorization_from(response, action),
+          test: test?,
+          error_code: error_code_from(response)
+        )
+      end
+
+      def parse(body)
+        return {} if body == ''
+
+        JSON.parse(body)
+      end
+
+      def message_from(response)
+        success_from(response) ? 'Transaction successful' : response['longText']
+      end
+
+      def error_code_from(response)
+        return unless success_from(response)
+
+        STANDARD_ERROR_CODE_MAPPING[response['primaryCode']]
+      end
+
+      def authorization_from(response, action)
+        ''
+      end
+
+      def request_headers(options)
+        headers = {
+          'Content-Type' => 'application/x-www-form-urlencoded'
+        }
+        headers['CompanyName'] = options[:company_name]
+        headers['AccessToken'] = options[:access_token] if options[:access_token].present?
+        headers['Invoice'] = options[:invoice] if options[:invoice].present?
+        headers
+      end
+
+      def success_from(response)
```

Comment:
updated this method as per the standards

Reply:
@ajawadmirza This method doesn't accurately check for errors in the response. Shift4 returns something like this with `error` inside an array.
```
@params=
  {"result"=>
    [{"error"=>
       {"longText"=>"No GTV PAN returned from host TOKEN01CE ENGINE16CE",
        "primaryCode"=>9101,
        "secondaryCode"=>6,
        "shortText"=>"SYSTEM ERROR"},
      "server"=>{"name"=>"UTGAPI09CE"}}]},
 @success=false,
 @test=true>
```

When the above is returned `response['error'].nil?` is always true and we add `'Transaction successful'` to the message even though it failed. Instead it should be `response['result'][0]['error'].nil?`. I don't think there's a standard way to write this method but happy to hear suggestions.

Reply:
Yes this should definitely be `response['result'.first['error'].nil?`. I only converted the same logic that was all in commit method to these standard methods `success_from(response)`. 

@ali-hassan  please make the said changes along with test cases.